### PR TITLE
Network policies

### LINKED ISSUE #106: [chart/redis_ha][REQUEST] Small improvements
I have a couple of small requests, most of which I could make a PR myself on, but I'd like to hear from maintainers first. :)

1. Our K8s cluster requires containers to define a `readinessProbe`. None of the containers in this chart do that right now. The readiness probe could just be the same `ping` that the `livenessProbe` does. I had to fix that on my fork and I'd love to make a PR on that.

2. Network policy - While those are not explicitly enforced they are mandated for our production environment. I can of course create one outside of this chart, but it'd be nice to have it as part of it. The chart could define a label that pods allowed to connect Redis/HAProxy must have in order to be allowed. I am however unsure exactly how the HAProxy <-> Redis connectivity works so not 100% on how to define the policy correctly (In our deployment we skipped HAProxy).

3. I have noticed that the cluster is in an incorrect state if you configure it with:
```yaml
redis:
    tlsReplication: true
    tlsPort: 6380
    port: 6379
  sentinel:
    tlsReplication: true
    tlsPort: 26380
    port: 26379
    auth: true
    existingSecret: <redacted>
    authKey: redis-password
    config:
      down-after-milliseconds: 5000
      failover-timeout: 10000
      parallel-syncs: 5
      maxclients: 10000
  auth: true
  existingSecret: <redacted>
  authKey: redis-password
  tls:
    secretName: <redacted>
    certFile: redis-crt
    keyFile: redis-key
    caCertFile: root-ca
```

The problem occurs when both `port` and `tlsPort` are defined. The Sentinels attempt to communicate with each other on the non-TLS port and that does not succeed. After `down-after-milliseconds` time all Sentinels assume the other Sentinels are gone. Thus they can never reach a quorum and perform a successful failover. After I reconfigured with:

```yaml
  redis:
    tlsReplication: true
    tlsPort: 6380
    port: 0
  sentinel:
    tlsReplication: true
    tlsPort: 26380
    port: 0
```

Things started working properly. I'm not sure what is the intention here, but at least some documentation warning that such misconfiguration is possible would be nice.

### PR DESCRIPTION
**Pull request by haljin**
#### Which issue this PR fixes
Closes #106 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [X] Chart Version bumped
- [X] Variables are documented in the README.md

### CODE REVIEW

**Review by DandyDeveloper — CHANGES_REQUESTED** [review]

File: charts/redis-ha/templates/redis-ha-network-policy.yaml [hunk 1]
```diff
@@ -0,0 +1,108 @@
+
+{{- if .Values.networkPolicy.enabled }}
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: {{ template "redis-ha.fullname" . }}-pdb
```
[the remark this answers is not in the document]
Comment (haljin) [reply charts/redis-ha/templates/redis-ha-network-policy.yaml:6]: Oops, completely missed that

**Comment by jdheyburn** [comment]
Hey, this looks like a useful PR. Is it possible for it to be merged in after a rebase?

File: charts/redis-ha/README.md [hunk 2]
```diff
@@ -210,6 +210,9 @@ The following table lists the configurable parameters of the Redis chart and the
 | `restore.s3.region`       | Restore init container - AWS AWS_REGION to access restore.s3.source                                                                                                                                      |``|
 | `restore.ssh.source`      | Restore init container - SSH scp location of dump - i.e. user@server:/path/dump.rdb                                                                                                                      | `false`                                                                                    |
 | `restore.ssh.key`         | Restore init container - SSH private key to scp restore.ssh.source to init container. Key should be in one line separated with \n. i.e. -----BEGIN RSA PRIVATE KEY-----\n...\n...\n-----END RSA PRIVATE KEY----- |``                                                                                 |
+| `networkPolicy.enabled`   | Enable generation of a NetworkPolicy  | `false` |
+| `networkPolicy.clientLabels`  | Pod labels for pods that should be allowed to connect to the Redis deployment | `[]` (everyone can connect) |
+| `networkPolicy.kubeDNSLabels` | Pod labels for Kube DNS | `[]` |
```
Comment (jdheyburn) [inline charts/redis-ha/README.md:215]: This should be updated to reflect the new values: `clientSelector` and `dnsSelector`.

File: charts/redis-ha/templates/redis-ha-network-policy.yaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+{{- if .Values.sentinel.tlsPort }}
+        - port: {{ .Values.sentinel.tlsPort }}
+          protocol: TCP
+{{- end }}   
+    - to: 
+{{ toYaml .Values.networkPolicy.dnsSelector | indent 8 }}
+      ports:
+        - port: 53
+          protocol: UDP
+{{- end }}
```
Comment (jdheyburn) [inline charts/redis-ha/templates/redis-ha-network-policy.yaml:92]: New line required here

**Comment by DandyDeveloper** [comment]
@haljin @jdheyburn Yes, I agree. I think the structure here is too rigid. 

I like the linked template. We should use that. 

If you have the time, can you create another PR @jdheyburn? I think this one may have gone stale.

**Comment by DandyDeveloper** [comment]
This also closes: https://github.com/DandyDeveloper/charts/issues/83

**Comment by sean-mcgimpsey** [comment]
Any update here ? We're really keen to implement network policies to isolate our redis statefulsets

**Comment by jdheyburn** [comment]
I've raised #146 which aims to introduce a minimal, but flexible Network Policy for Redis Pods. Please take a look and continue the discussion there.

@sean-mcgimpsey @DandyDeveloper

**Comment by haljin** [comment]
I guess I will abandon this then, thanks @jdheyburn !

**Pull request closed without merging**